File: Aion2FunDps.Protocol/CompressionDetector.cs
namespace Aion2FunDps.Protocol;

/// <summary>
/// Detects whether a game packet body is LZ4-compressed via the 0xff 0xff marker.
/// Some packets carry "extra flag" bytes (range 0xf0..0xfe) before the marker.
///
/// Phase 1b only DETECTS. Actual LZ4 decompression is Phase 1c.
/// </summary>
public static class CompressionDetector
{
    public static bool IsCompressed(ReadOnlySpan<byte> body, out int compressedDataOffset)
    {
        compressedDataOffset = 0;
        int i = 0;

        // Skip extra-flag bytes (0xf0..0xfe).
        while (i < body.Length && body[i] >= 0xf0 && body[i] <= 0xfe)
            i++;

        // Look for 0xff 0xff marker.
        if (i + 1 < body.Length && body[i] == 0xff && body[i + 1] == 0xff)
        {
            compressedDataOffset = i + 2;
            return true;
        }
        return false;
    }
}

File: Aion2FunDps.Engine/src/compression_detector.cpp
// compression_detector.cpp — implementation.

#include "compression_detector.h"

namespace aion2fun {

CompressionProbe is_compressed(const uint8_t* body, size_t length) noexcept {
    CompressionProbe probe{0, false};
    if (body == nullptr || length < 2) return probe;

    // Skip "extra-flag" bytes in the 0xf0..0xfe range. Some packets
    // carry one or more of these before the actual marker pair; the
    // semantics of the flags themselves aren't relevant to compression
    // — we just step past them.
    size_t i = 0;
    while (i < length && body[i] >= 0xf0 && body[i] <= 0xfe) {
        ++i;
    }

    // Look for the 0xff 0xff pair.
    if (i + 1 < length && body[i] == 0xff && body[i + 1] == 0xff) {
        probe.compressed_data_offset = static_cast<int32_t>(i + 2);
        probe.is_compressed = true;
    }
    return probe;
}

}  // namespace aion2fun
